joomla/joomla-cms
# [4.0] Add back remote database host security check at installation

Pull Request for Issue #18736 .

### Summary of Changes

This pull request (PR) adds back the special security check for "remote" database hosts at new installation which works in J3 and is described here: [https://docs.joomla.org/Special:MyLanguage/J3.x:Secured_procedure_for_installing_Joomla_with_a_remote_database](https://docs.joomla.org/Special:MyLanguage/J3.x:Secured_procedure_for_installing_Joomla_with_a_remote_database).

 "remote" in quotes because this check also runs if a valid IP address or host name (short or FQDN) is given for a local database server (local means database server and web server = database client on the same host). The check assumes anything else than "localhost", "127.0.0.1" and "::1" to be a remote host. I think this was desired because as soon as forgotten Joomla installation is reachable from outside, it needs this security check.

### Testing Instructions

Pre-condition: You either have a remote database server which can be accessed with a database client from your webserver, or you have both at the same host and set up your database sevrer so it listens to an IP address different to "127.0.0.1" (v4) or "::1" (v6), and optionally have a real host name different to "localhost" which can be resolved to such an IP address. If you have a remote database server you should also have a local one to test at the end that there is no security check it not necessary.

1. Make sure your webserver has read and write access to the installation folder (should be the case for a clean git branch or an unzipped nightly build).

2. Start a new installation of a 4.0-dev branch plus changes of this PR applied, i.e. either apply it with some git tool on a clean 4.0-dev branch, or apply it on an existing 4.0-dev installation e.g. with patchtester and then remove configuration.php and delete the database tables.

3. Fill in all necessary data in the installation form and use something different to "localhost" and "127.0.0.1" and "::1" (see pre-conditions above).

4. Click the "Install Joomla" button at the bottom.
Result: See screenshot 1 in section "Expected result" below. The installation does not continue. An alert tells to delete a file created by Joomla.

5. Delete the file as advised in the error alert.

6. Click the "Install Joomla" button at the bottom.
Result: See screenshot section "Actual result" below. The installation finish with success.

7. Remove configuration.php, delete the database tables and delete the session cookie to clear session data.

8. Make sure your webserver doesn't have write permission for the "installation" folder, e.g. by entering following in a Linux command shell as the user who is owner of the "installation" folder and with current directory = Joomla root: `chmod -w installation`. Or on Windows, add a permission which forbids anybody to write.

9. Start again a new installation.

10. Fill in all necessary data in the installation form and use something different to "localhost" and "127.0.0.1" and "::1" (see pre-conditions above).

11. Click the "Install Joomla" button at the bottom.
Result: See screenshot 2 in section "Expected result" below. The installation does not continue. An alert tells to upload or create a file with a given name.

12. Change back permissions to like they were before step 8, e.g. on Linux: `chmod +w installation`.

13. Upload or create the file as advised in the error alert.

14. Click the "Install Joomla" button at the bottom.
Result: See screenshot section "Actual result" below. The installation finish with success.

15. Remove configuration.php, delete the database tables and delete the session cookie to clear session data.

16. Start again a new installation.

17. Fill in all necessary data in the installation form and this time use either "localhost" or "127.0.0.1" or "::1" (see pre-conditions above).

18. Click the "Install Joomla" button at the bottom.
Result: See screenshot section "Actual result" below. The installation finish with success without the security check for remote hosts.

### Expected result

1. If the file could be created by Joomla Installation:
![new-01](https://user-images.githubusercontent.com/7413183/71558472-1bc90600-2a54-11ea-8a74-dcc0247368d1.png)

2. If the file couldn't be created by Joomla Installation:
![new-02](https://user-images.githubusercontent.com/7413183/71558502-867a4180-2a54-11ea-993a-d94fa7d48bf6.png)

3. After all instructions shown in the alert have been completed: See actual result below.

### Actual result

![snap-02](https://user-images.githubusercontent.com/7413183/71548034-7ad14100-29a8-11ea-9621-ed82ce034bb7.png)

### Documentation Changes Required

None I think.
